### PR TITLE
Add new compiler.get_argument_syntax method

### PR DESCRIPTION
**Pull request by dcbaker**
Some compilers try very had to pretend they're another compilers (ICC
pretends to be GCC and Linux and MacOS, and MSVC on windows), Clang
behaves much like GCC, but now also has an MSVC like build.

This method provides a way to determine what the compiler behaves like,
which is useful for differentiating two compilers that have the same
name but take different arguments on different platforms, as well as for
grouping compiler checks

### CODE REVIEW

**Comment by dcbaker** [comment]
@jon-turney, have I missed any corner cases you can think of?

File: mesonbuild/compilers/compilers.py [hunk 1]
```diff
@@ -1398,6 +1409,8 @@ def openmp_flags(self):
 
     def gnu_symbol_visibility_args(self, vistype):
         return gnu_symbol_visibility_args[vistype]
+    def get_argument_type(self):
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/compilers.py]: __[Flake8]__

[E301] expected 1 blank line, found 0

**Comment by dcbaker** [comment]
I should probably also mention that this should simplify adding ICL (ICC for windows) and clang-cl support, which is why I wrote this.

File: docs/markdown/Compiler-properties.md [hunk 2]
```diff
@@ -29,9 +29,14 @@ Compiler id
...
 
 The compiler object has a method called `get_id`, which returns a
-lower case string describing the "family" of the compiler. See
-[reference tables](Reference-tables.md) for a list of supported
-compiler ids.
+lower case string describing the "family" of the compiler.
+
+The compiler object also has a method `get_argument_type` which 
+returns a lower case string of `gcc`, `msvc`, or `other`; identifying
+that the compiler takes the same arguments as either `gcc` or `msvc`.
```
Comment (jon-turney) [inline docs/markdown/Compiler-properties.md]: "same style arguments" would make more sense than "same arguments", but I'm not sure if we actually need to say anything here, because...

File: docs/markdown/Reference-manual.md [hunk 3]
```diff
@@ -1683,6 +1683,12 @@ the following methods:
 - `get_id()` returns a string identifying the compiler. For example,
   `gcc`, `msvc`, [and more](Reference-tables.html#compiler-ids).
 
+- `get_argument_type()` *(new in 0.49.0) returns a string identifying the type
+  of arguments the compiler takes. Can be one of `gcc`, `msvc`, or `other`.
+  This method is useful for identifying compilers that are not gcc or msvc,
+  but accept the same kinds of arguments as one of those two compilers such as
+  clang or icc.
```
Comment (jon-turney) [inline docs/markdown/Reference-manual.md]: Can we just leave it undefined what it says in other cases, rather than explicitly saying it must say "other"?

I wonder if we should say something like "if you need to use this outside of the meson test suite, that probably indicates a missing feature in meson" here?

File: docs/markdown/snippets/compiler_argument_type.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  if cc.has_argument('/w1235')
+    add_project_arguments('/w1235', language : ['c'])
+  endif
+elif cc.get_argument_type() == 'gcc'
+  if cc.has_argument('-Wfoo')
+    add_project_arguments('-Wfoo', language : ['c'])
+  endif
+else
+  if cc.has_argument('--error-on-foo')
+    add_project_arguments('--error-on-foo', language : ['c'])
```
Comment (jon-turney) [inline docs/markdown/snippets/compiler_argument_type.md]: I'm not sure what kind of compiler this 'else' case is meant to apply to.

I'm not sure this example should do anything but either ignore, or raise an error in this case, since those are the only things it's safe to do.

File: test cases/common/205 argumnet type/meson.build [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+elif cc.get_id() == 'msvc'
+  expected = 'msvc'
+elif cc.get_id() == 'icc'
+  if host_machine.system() == 'windows'
+    expected = 'msvc'
+  else
+    expected = 'gcc'
+  endif
+else
+  # It's possible that other compilers end up here that shouldn't
```
Comment (jon-turney) [inline test cases/common/205 argumnet type/meson.build]: Probably the test should fail here: It indicates a compiler the test case doesn't know about?

File: test cases/common/205 argumnet type/meson.build [hunk 6]
```diff
@@ -0,0 +1,25 @@
+project(
```
Comment (jon-turney) [inline test cases/common/205 argumnet type/meson.build]: 'argumnet' ?

**Comment by jon-turney** [comment]
>  have I missed any corner cases you can think of?

Looks complete to me

File: docs/markdown/Reference-manual.md (hunk 3 above)
Comment (dcbaker) [reply docs/markdown/Reference-manual.md]: I'm pretty certain for as we add more GCC and MSVC like compilers people are going to want this in their build system, since it behaves like GCC if the build machine is MacOS or Linux, or like MSVC if the build machine is Windows.
```meson
cc = meson.get_compiler('c')
if ['clang-cl', 'msvc', 'wacom'].contains(cc.get_id()) or (cc.get_id() == 'intel' and build_machine.system() == 'windows')
```
or:
```meson
cc = meson.get_compiler('c')
if cc.get_argument_type() == 'msvc'
```

File: test cases/common/205 argumnet type/meson.build (hunk 6 above)
Comment (dcbaker) [reply test cases/common/205 argumnet type/meson.build]: oops :)

File: docs/markdown/snippets/compiler_argument_type.md (hunk 4 above)
Comment (jon-turney) [reply docs/markdown/snippets/compiler_argument_type.md]: I'm just saying this example could set a better example.  What compiler supports '--error-on-foo'? Should you check get_id() for that, rather than assuming that *all* compilers of an unknown argument type support it?

File: docs/markdown/Reference-manual.md (hunk 3 above)
Comment (jon-turney) [reply docs/markdown/Reference-manual.md]: You've omitted the body of the `if`, which presumably contains something that meson lacks the means to express more portably.  So I think we are in agreement here :)

**Comment by jon-turney** [comment]
I think we need to be a bit careful about what the information this method returns is actually supposed to mean.

Internally, and in test cases, we check if the compiler id is 'msvc' to check:
* if the C compiler can directly compile assembly
* if pch requires a source file
* the library naming convention to use
* if -D options ending in '\' or '\"' require strange quoting (???)
* the argument style the compiler accepts
* probably something else I've overlooked

I think there will be a temptation will be to use this method to check for any and all of these things.  So maybe this needs to be named something like `get_group_id()`.  Or maybe we need to handle each of thse details individually.

(Some of these things don't make sense to surface in the meson.build e.g. pch style; that test should probably become a compiler class method)

(The assembly issue seems like a missing abstraction in meson.  It could use the appropriate tool (gcc/gas/masm) to process .s/.S files, depending on what's available?)

**Comment by nirbheek** [comment]
> (The assembly issue seems like a missing abstraction in meson. It could use the appropriate tool (gcc/gas/masm) to process .s/.S files, depending on what's available?)

Yes, for MSVC we should be using `cl /EP /P` to pre-process and `ml.exe` to compile. I just haven't gotten around to implementing it yet. I need it for libffi.

**Comment by dcbaker** [comment]
I think the things that this is useful for are:
- argument style the compiler checks
- The library naming convention

Things it may be useful for:
- If the c compiler can directly compile assembly (ICC on windows doesn't support assembly, ICC on Linux and MacOS does, at least. Not sure about clang-cl)

Should we add this to the documentation?

**Comment by jon-turney** [comment]
> I think the things that this is useful for are:
> 
>     * argument style the compiler checks
> 
>     * The library naming convention
> 
> [...]
> 
> Should we add this to the documentation?

My main concern in case (a) above (which seems like the sensible one) was the name, but yes.

**Comment by jpakkane** [comment]
Rather than talk about argument types, maybe a better syntax would be argument _format_? A "type" has a specific meaning in languages and this is not really it.

Also, a design mistake I wish I did not make and could fix was adding `get_` in front of methods like these. Unfortunately I fear that ship has sailed, sunk and become an immovable coral reef that can't ever be changed. But it still annoys me every time.

**Comment by dcbaker** [comment]
I can change the name, do you like `get_argument_format` or `get_argument_syntax` better?

**Comment by jon-turney** [comment]
> Actually, the more I think about it, the more I think that only the argument style is really the right thing to check with this. even the naming convention isn't really tied, since MinGW can be used to create .dll's and it's not at all like MSVC, right?

'library naming convention' means the mess of stuff that `CCompiler.get_library_naming()` does, not 'the shared library extension'.  All targets which use .dll shared libraries are not equivalent in that regard.

Please take a look at the changes to test cases in #4250.  With your proposed approach, there are many places where I need to make a choice between:
* `if meson.get_compiler('c').get_argument_type() == 'msvc'` 
* `if ['msvc', 'clang-cl'].contains(meson.get_compiler('c').get_id())`

and you'll need to come along and change probably all of the latter again when you add ICC/Windows.

**Comment by dcbaker** [comment]
ping

**Comment by dcbaker** [comment]
Fixed typos, reworded a couple of things in the docs to (hopefully) make them clearer, and moved the test case to 206.

File: docs/markdown/Compiler-properties.md [hunk 7]
```diff
@@ -29,9 +29,16 @@ Compiler id
...
-lower case string describing the "family" of the compiler. See
-[reference tables](Reference-tables.md) for a list of supported
-compiler ids.
+lower case string describing the "family" of the compiler.
+
+The compiler object also has a method `get_argument_syntax` which 
+returns a lower case string of `gcc`, `msvc`, or another undefined string
+value; identifying whether the compiler arguments use the same syntax as
+either `gcc` or `msvc`, or that it's arguments are not like either. This should
+only be used to select the syntax of the arguments to test.
```
Comment (jon-turney) [inline docs/markdown/Compiler-properties.md]: 'to test'? ... with `has_argument()` presumably?

I like this, but I'm not sure we should be quite that restrictive:  There are instances in test cases where we (1) add arguments we 'know' are valid for msvc, or (2) build compilation commands entire.

File: docs/markdown/Reference-tables.md [hunk 8]
```diff
@@ -4,26 +4,26 @@
 
 These are return values of the `get_id` method in a compiler object.
```
Comment (jon-turney) [inline docs/markdown/Reference-tables.md]: The sentence prefacing the table doesn't make sense after you've added a column.  Do you really want to say 'other' when the value of get_argument_syntax is undefined.  Would it make more sense to have a separate table for this?

Comment (dcbaker) [reply docs/markdown/Reference-tables.md]: I tried a separate table, I didn't like that the keys of the table were copied between the two tables. I've removed "other" now undefined values are just left blank.

File: docs/markdown/Compiler-properties.md (hunk 7 above)
Comment (dcbaker) [reply docs/markdown/Compiler-properties.md]: I've tried to clarify and soften that a little bit.

**Comment by dcbaker** [comment]
@jon-turney, @jpakkane, ping

**Comment by jpakkane** [comment]
> or that it's arguments

That should be "its".

 - "it's" means _only_ either "it is" or "it has" 
 - "its" is the possessive form of "it"

Other than that LGTM. @jon-turney?

**Comment by dcbaker** [comment]
gah. I'm a native english speaker and I still find apostrophes and pronouns confusing :)

**Comment by jon-turney** [comment]
I have nothing further to add.
